text/0000-api-version-compat.md

```diff
@@ -0,0 +1,119 @@
+- purely additive features
+- simple flag/file/key renames (the previous usage will be supported but deprecated by the lifecycle)
+- removal of features (these will be interpreted as deprecated by the lifecycle)
+
+## Lifecycle
+The lifecycle will continue use the [lifecycle descriptor](https://github.com/buildpacks/rfcs/blob/pack-publish-buildpack/text/0011-lifecycle-descriptor.md) 
+to indicate the implemented API versions as before. But the assumptions we make about compatibility will change.
+
+## Deprecations
+If a feature is removed from the lifecycle's advertised API version but used by a platform or buildpack the lifecycle will print a deprecation warning.
+Because platform authors may want to shield end users from deprecation warning caused by the platform's use of deprecated
+lifecycle features, the lifecycle will provide the following options to configure deprecation warnings:
+
+All lifecycle binaries will have a deprecation mode, configurable in the following way:
+| Flag              | Environment Variable | Default
+|-------------------|----------------------|--------
+| -deprecation-mode | CNB_DEPRECATION_MODE | all
+
+The allowed values of deprecation mode are as follows:
+| Mode      | Behavior
+|-----------|---------
+| all       | prints all deprecation warnings
+| buildpack | prints only buildpack API deprecation warnings
+| platform  | prints only platform API deprecation warnings
+| none      | silences all deprecation warnings
+
+# How it Works
+[how-it-works]: #how-it-works
+## Spec
+The core team will ensure that no changes are added to a minor spec release that imply incompatibility with the previous minor.
+
+Examples of ALLOWED changes in a minor 0.x API version:
+1. the addition of new image labels
+1. the addition of a key to an existing label
+1. the addition of a key to an existing file
+1. rename a file or a key in a file supplied as input
+1. rename a flag
+1. remove a flag
+
+Example of DISALLOWED changes in a minor API version:
+1. changes to the contractual build plan
+2. type changes to an existing key in a label or output file
+
+## Lifecycle
+The lifecycle will print deprecation warnings according to the rules described above. The lifecycle will
+guarantee that all usage that conforms to API `0.n` will continue to work, even when indicating API `0.n+1` is implemented.
+
+## Platforms
+Platforms that wish to enforce compatibility should now treat all minor API versions as non-breaking with the exception
+of buildpack and platform `0.1` -> `0.2`. These API versions did include breaking changes and will not be changed.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The spec and lifecycle now carry the burden of ensuring there are no breaking changes before `1.0`. `1.0` is an important symbolic
+milestone and there could be unforseen and compelling reasons to make a breaking change before we are ready to release `1.0` APIs.
```

Comment:
With all the [potential breaking changes for 1.0](https://docs.google.com/document/d/1ZETueW4JoumOfqa37VICKwcOgEKM2juYtNI9rRv0McM/edit) already outlined, can we achieve this by just doing minor changes and making 1.0 a non event?

Reply:
Probably! This RFC doesn't state that `1.0` must be breaking, it keeps the status quo - it can be breaking if it wishes.

Reply:
Even if a `1.0` API version contains breaking changes the `1.0` release of the lifecycle doesn't have to be breaking. I have some idea around facilitating a smooth lifecycle `1.0` upgrade experience but I want to keep them out of scope for this RFC which is focusing on the more immediate problem - future 0.x API versions.